### PR TITLE
[ie/zetland] Add ZetlandDKArticle Extractor

### LINKED ISSUE #9024: Zetland.dk
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

DK

### Example URLs

https://www.zetland.dk/historie/sO9aq2MY-a81VP3BY-66e69?utm_source=instagram&utm_medium=linkibio&utm_campaign=artikel

### Provide a description that is worded well enough to be understood

Please add support for zetland.dk

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://www.zetland.dk/historie/sO9aq2MY-a81VP3BY-66e69?utm_source=instagram&utm_medium=linkibio&utm_campaign=artikel']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version nightly@2024.01.14.232710 from yt-dlp/yt-dlp-nightly-builds [8e6e36517] (zip)
[debug] Python 3.10.12 (CPython x86_64 64bit) - Linux-6.2.0-1018-lowlatency-x86_64-with-glibc2.35 (OpenSSL 3.0.2 15 Mar 2022, glibc 2.35)
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2
[debug] Optional libraries: Cryptodome-3.11.0, certifi-2023.11.17, pyxattr-0.7.2, requests-2.25.1, secretstorage-3.3.1, sqlite3-3.37.2, urllib3-1.26.5
[debug] Proxy map: {}
[debug] Request Handlers: urllib
[debug] Loaded 1798 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-nightly-builds/releases/latest
Latest version: nightly@2024.01.14.232710 from yt-dlp/yt-dlp-nightly-builds
yt-dlp is up to date (nightly@2024.01.14.232710 from yt-dlp/yt-dlp-nightly-builds)
[generic] Extracting URL: https://www.zetland.dk/historie/sO9aq2MY-a81VP3BY-66e69?utm_source=instagram&utm_medium=linkibio&utm_campaign=artikel
[generic] sO9aq2MY-a81VP3BY-66e69?utm_source=instagram&utm_medium=linkibio&utm_campaign=artikel: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] sO9aq2MY-a81VP3BY-66e69?utm_source=instagram&utm_medium=linkibio&utm_campaign=artikel: Extracting information
[debug] Looking for embeds
ERROR: Unsupported URL: https://www.zetland.dk/historie/sO9aq2MY-a81VP3BY-66e69?utm_source=instagram&utm_medium=linkibio&utm_campaign=artikel
Traceback (most recent call last):
  File "/usr/local/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1587, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/local/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1722, in __extract_info
    ie_result = ie.extract(url)
  File "/usr/local/bin/yt-dlp/yt_dlp/extractor/common.py", line 718, in extract
    ie_result = self._real_extract(url)
  File "/usr/local/bin/yt-dlp/yt_dlp/extractor/generic.py", line 2531, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.zetland.dk/historie/sO9aq2MY-a81VP3BY-66e69?utm_source=instagram&utm_medium=linkibio&utm_campaign=artikel
```

### PR DESCRIPTION
**Pull request by HobbyistDev**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

This PR add site support for `zetland.dk`. This PR only test the OP url in #9024 as the other url require subscription. This PR only assume the article only contain 1 audio link with no other type of media

Fixes #9024

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/zetland.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    def _real_extract(self, url):
+        display_id, uploader_id = self._match_valid_url(url).group('id', 'uploader_id')
+        webpage = self._download_webpage(url, display_id)
+
+        next_js_data = self._search_nextjs_data(webpage, display_id)['props']['pageProps']
+        story_data = traverse_obj(next_js_data, ('initialState', 'consume', 'story', 'story'))
+
+        formats = []
+        for audio_url in traverse_obj(story_data, ('story_content', 'meta', 'audioFiles', ...)):
```
Comment (bashonly) [inline yt_dlp/extractor/zetland.py:38]: would be best to ensure URLs with a `url_or_none` transformation
```suggestion
        for audio_url in traverse_obj(story_data, ('story_content', 'meta', 'audioFiles', ..., {url_or_none})):
```

File: yt_dlp/extractor/zetland.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            formats.append({
+                'url': audio_url,
+                'vcodec': 'none',
+                'ext': determine_ext(audio_url)
```
Comment (bashonly) [inline yt_dlp/extractor/zetland.py:42]: core code will set `ext` from `determine_ext(url)` if the extractor doesn't set it, so it's better to keep it simple imo
```suggestion
                'url': audio_url,
                'vcodec': 'none',
```

File: yt_dlp/extractor/zetland.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            'uploader_id': uploader_id
+        }, traverse_obj(story_data, {
+            'title': ('story_content', 'content', 'title') or 'title',
+            'uploader': ('sharer', 'name'),
+            'uploader_id': ('sharer', 'sharer_id'),
+            'description': ('story_content', 'content', 'sosialDescription'),
+            'series_id': ('story_content', 'meta', 'seriesId'),
+            'release_timestamp': ('published_at', {unified_timestamp}),
+            'modified_timestamp': ('revised_at', {unified_timestamp}),
+        }), traverse_obj(next_js_data, ('metaInfo', {
```
Comment (bashonly) [inline yt_dlp/extractor/zetland.py:57]: `or` won't work as you want it to within a traversal path, you need to use branching instead, with `get_all=False` so it doesn't return a list
```suggestion
        }, traverse_obj(story_data, {
            'title': ((('story_content', 'content', 'title'), 'title'), {str}),
            'uploader': ('sharer', 'name'),
            'uploader_id': ('sharer', 'sharer_id'),
            'description': ('story_content', 'content', 'sosialDescription'),
            'series_id': ('story_content', 'meta', 'seriesId'),
            'release_timestamp': ('published_at', {unified_timestamp}),
            'modified_timestamp': ('revised_at', {unified_timestamp}),
        }, get_all=False), traverse_obj(next_js_data, ('metaInfo', {
```

File: yt_dlp/extractor/zetland.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            'title': ('meta', 'title') or ('ld', 'headline') or ('og', 'og:title') or ('og', 'twitter:title'),
+            'description': (('meta', 'description') or ('ld', 'description')
+                            or ('og', 'og:description') or ('og', 'twitter:description')),
+            'uploader': ('meta', 'author') or ('ld', 'author', 'name'),
+            'uploader_url': ('ld', 'author', 'url'),
+            'thumbnail': ('ld', 'image') or ('og', 'og:image') or ('og', 'twitter:image'),
+            'modified_timestamp': ('ld', 'dateModified', {unified_timestamp}),
+            'release_timestamp': ('ld', 'datePublished', {unified_timestamp}),
+            'timestamp': ('ld', 'dateCreated', {unified_timestamp}),
+        })), {
```
Comment (bashonly) [inline yt_dlp/extractor/zetland.py:67]: same as above
```suggestion
            'title': ((('meta', 'title'), ('ld', 'headline'), ('og', 'og:title'), ('og', 'twitter:title')), {str}),
            'description': ((('meta', 'description'), ('ld', 'description'), ('og', 'og:description'), ('og', 'twitter:description')), {str}),
            'uploader': ((('meta', 'author'), ('ld', 'author', 'name')), {str}),
            'uploader_url': ('ld', 'author', 'url', {url_or_none}),
            'thumbnail': ((('ld', 'image'), ('og', 'og:image'), ('og', 'twitter:image')), {url_or_none}),
            'modified_timestamp': ('ld', 'dateModified', {unified_timestamp}),
            'release_timestamp': ('ld', 'datePublished', {unified_timestamp}),
            'timestamp': ('ld', 'dateCreated', {unified_timestamp}),
        }), get_all=False), {
```

File: yt_dlp/extractor/zetland.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            'title': self._html_search_meta(['title', 'og:title', 'twitter:title'], webpage),
+            'description': self._html_search_meta(['description', 'og:description', 'twitter:description'], webpage),
+            'thumbnail': self._html_search_meta(['og:image', 'twitter:image'], webpage),
+            'uploader': self._html_search_meta(['author'], webpage),
+            'release_timestamp': unified_timestamp(self._html_search_meta(['article:published_time'], webpage)),
+        }, self._search_json_ld(webpage, display_id))
```
Comment (bashonly) [inline yt_dlp/extractor/zetland.py:73]: json ld should be non-fatal since it is metadata-only and a fallback to a fallback

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/zetland.py (hunk 5 above)
Comment (HobbyistDev) [reply yt_dlp/extractor/zetland.py]: Resolved in b9aa2b560470266f015de026f4c8ecbbb9591d52

File: yt_dlp/extractor/zetland.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            'modified_date': '20240118',
+            'title': 'Afsnit 1: “Det føltes som en kidnapning.” ',
+            'upload_date': '20240116',
+            'uploader_id': 'a81VP3BY',
+            'modified_timestamp': 1705568739,
+            'release_timestamp': 1705377592,
+            'uploader_url': 'https://www.zetland.dk/skribent/a81VP3BY',
+            'uploader': 'Helle Fuusager',
+            'release_date': '20240116',
+            'thumbnail': 'https://zetland.imgix.net/2aafe500-b14e-11ee-bf83-65d5e1283a57/Zetland_Image_1.jpg?fit=crop&crop=focalpoint&auto=format,compress&cs=srgb&fp-x=0.49421296296296297&fp-y=0.48518518518518516&w=1200&h=630',
```
Comment (bashonly) [inline yt_dlp/extractor/zetland.py:22]: use regex here imo
```suggestion
            'thumbnail': r're:https://zetland\.imgix\.net/2aafe500-b14e-11ee-bf83-65d5e1283a57/Zetland_Image_1\.jpg',
```

File: yt_dlp/extractor/zetland.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
+from .common import InfoExtractor
+from ..utils import merge_dicts, unified_timestamp, url_or_none
+from ..utils.traversal import traverse_obj
+
+
+class ZetlandDKArticleIE(InfoExtractor):
+    _VALID_URL = r'https://www.zetland.dk/\w+/(?P<id>(?P<story_id>\w{8})-(?P<uploader_id>\w{8})-(?:\w{5}))'
```
Comment (bashonly) [inline yt_dlp/extractor/zetland.py:7]:
```suggestion
    _VALID_URL = r'https://www\.zetland\.dk/\w+/(?P<id>(?P<story_id>\w{8})-(?P<uploader_id>\w{8})-(?:\w{5}))'
```

**Review by bashonly — APPROVED** [review]

**Comment by bashonly** [comment]
please double check the test since the branching fixes may yield different results

**Comment by HobbyistDev** [comment]
The test looks fine to me:
```shell
> python devscripts/run_tests.py ZetlandDKArticleIE
Running ['pytest', '-Werror', '--tb=short', 'test/test_download.py::TestDownload::test_ZetlandDKArticle']
Running ['C:\\Users\\HobbyistDev\\Documents\\Program\\python_proj\\Scripts\\python.exe', '-Werror', '-m', 'unittest', 'test.test_download.TestDownload.test_ZetlandDKArticle']
[debug] Loaded 1825 extractors 
[ZetlandDKArticle] Extracting URL: https://www.zetland.dk/historie/sO9aq2MY-a81VP3BY-66e69?utm_source=instagram&utm_medium=linkibio&utm_campaign=artikel
[ZetlandDKArticle] sO9aq2MY-a81VP3BY-66e69: Downloading webpage
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] sO9aq2MY-a81VP3BY-66e69: Downloading 1 format(s): 2
[info] Writing video metadata as JSON to: test_ZetlandDKArticle_sO9aq2MY-a81VP3BY-66e69.info.json 
[debug] Invoking http downloader on "https://storage.googleapis.com/zetland-upload-production/07134420-b45f-11ee-b3a9-03589fc21889%2FJoannahusetafsnit1NYNYFINAL.mp3"
[debug] File locking is not supported. Proceeding without locking[download] Destination: test_ZetlandDKArticle_sO9aq2MY-a81VP3BY-66e69.mp3 
[download] 100% of   10.00KiB in 00:00:02 at 4.65KiB/s 
.
----------------------------------------------------------------------
Ran 1 test in 4.214s

OK
```

File: yt_dlp/extractor/zetland.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
+from .common import InfoExtractor
+from ..utils import merge_dicts, unified_timestamp, url_or_none
+from ..utils.traversal import traverse_obj
+
+
+class ZetlandDKArticleIE(InfoExtractor):
+    _VALID_URL = r'https://www\.zetland\.dk/\w+/(?P<id>(?P<story_id>\w{8})-(?P<uploader_id>\w{8})-(?:\w{5}))'
```
Comment (seproDev) [inline yt_dlp/extractor/zetland.py:7]:
```suggestion
    _VALID_URL = r'https?://www\.zetland\.dk/\w+/(?P<id>(?P<story_id>\w{8})-(?P<uploader_id>\w{8})-(?:\w{5}))'
```

**Review by seproDev — APPROVED** [review]

**Pull request merged**